src/main/java/jenkins/plugins/git/GitPublisherStep.java

```diff
@@ -0,0 +1,367 @@
+import org.jenkinsci.plugins.gitclient.GitClient;
+import org.jenkinsci.plugins.gitclient.PushCommand;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepDescriptorImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.kohsuke.stapler.AncestorInPath;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+/**
+ * Runs Git Publishing steps using {@link GitSCM}.
+ */
+public final class GitPublisherStep extends AbstractStepImpl {
+
+  private final String url;
+  private String branch = "master";
+  private String credentialsId;
+
+  private boolean forcePush;
+
+  private List<GitPublisher.TagToPush> tagsToPush;
+  // Pushes HEAD to these locations
+  private List<GitPublisher.BranchToPush> branchesToPush;
+  // notes support
+  private List<GitPublisher.NoteToPush> notesToPush;
+
+  @DataBoundConstructor
+  public GitPublisherStep(String url) {
+    this.url = url;
+  }
+
+  public String getUrl() {
+    return url;
+  }
+
+  public String getBranch() {
+    return branch;
+  }
+
+  public String getCredentialsId() {
+    return credentialsId;
+  }
+
+  @DataBoundSetter
+  public void setBranch(String branch) {
+    this.branch = branch;
+  }
+
+  @DataBoundSetter
+  public void setCredentialsId(String credentialsId) {
+    this.credentialsId = Util.fixEmpty(credentialsId);
+  }
+
+  @DataBoundSetter
+  public void setTagsToPush(List<GitPublisher.TagToPush> tagsToPush) {
+    this.tagsToPush = tagsToPush;
+  }
+
+  @DataBoundSetter
+  public void setBranchesToPush(List<GitPublisher.BranchToPush> branchesToPush) {
+    this.branchesToPush = branchesToPush;
+  }
+
+  @DataBoundSetter
+  public void setNotesToPush(List<GitPublisher.NoteToPush> notesToPush) {
+    this.notesToPush = notesToPush;
+  }
+
+  public SCM createSCM() {
+    return new GitSCM(GitSCM.createRepoList(url, credentialsId), Collections.singletonList(new BranchSpec("*/" + branch)), false, Collections.<SubmoduleConfig>emptyList(), null,
+      null, Collections.<GitSCMExtension>singletonList(new LocalBranch(branch)));
+  }
+
+  public boolean isForcePush() {
+    return forcePush;
+  }
+
+  public boolean isPushTags() {
+    if (tagsToPush == null) {
+      return false;
+    }
+    return !tagsToPush.isEmpty();
+  }
+
+  public boolean isPushBranches() {
+    if (branchesToPush == null) {
+      return false;
+    }
+    return !branchesToPush.isEmpty();
+  }
+
+  public boolean isPushNotes() {
+    if (notesToPush == null) {
+      return false;
+    }
+    return !notesToPush.isEmpty();
+  }
+
+  public List<GitPublisher.TagToPush> getTagsToPush() {
+    if (tagsToPush == null) {
+      tagsToPush = new ArrayList<GitPublisher.TagToPush>();
+    }
+
+    return tagsToPush;
```

Comment:
Almost 🐛 . IMHO all the lists should be returned as unmodifyable lists to prevent the exposure of the internal implementation